src/utils/src/arg_parser.rs

```diff
@@ -0,0 +1,567 @@
+        Ok(Help::NotProvided)
+    }
+
+    fn validate_matches(&self, args: &[String]) -> Result<()> {
+        for arg_info in self.args.values() {
+            // The arguments that are marked `required` must be provided by user.
+            if arg_info.required && arg_info.user_value.is_none() {
+                return Err(Error::MissingArgument(arg_info.name.to_string()));
+            }
+            // For the arguments that require a specific argument to be also present in the list
+            // of arguments provided by user, search for that argument.
+            if arg_info.user_value.is_some() {
+                if let Some(arg_name) = arg_info.requires {
+                    if !args.contains(&(format!("--{}", arg_name))) {
+                        return Err(Error::MissingArgument(arg_name.to_string()));
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn validate_arg(&self, arg: &str) -> Result<()> {
+        if !arg.starts_with(ARG_SEPARATOR) {
+            return Err(Error::UnexpectedArgument(arg.to_string()));
+        }
+        let arg_name = &arg[ARG_SEPARATOR.len()..];
+
+        // Check if the argument is an expected one.
+        if !self.args.contains_key(arg_name) {
+            return Err(Error::UnexpectedArgument(arg_name.to_string()));
+        }
+
+        // Check if the argument was not provided more than once.
+        if self.args.get(arg_name).unwrap().user_value.is_some() {
+            return Err(Error::DuplicateArgument(arg_name.to_string()));
+        }
+        Ok(())
+    }
+
+    // Populates the fields of ArgParser.
+    pub fn get_matches(&mut self, args: &[String]) -> Result<()> {
+        let mut iter = args.iter();
+        while let Some(arg) = iter.next() {
+            self.validate_arg(arg)?;
+
+            let arg_info = self
+                .args
+                .get_mut(&arg[ARG_SEPARATOR.len()..])
+                .ok_or_else(|| Error::UnexpectedArgument(arg[ARG_SEPARATOR.len()..].to_string()))?;
+
+            let val = if arg_info.takes_value {
+                iter.next()
+                    .filter(|v| !v.starts_with(ARG_SEPARATOR))
+                    .ok_or_else(|| Error::MissingValue(arg_info.name.to_string()))?
+            } else {
+                FLAG_PROVIDED
+            };
+
+            arg_info.user_value = Some(val.to_string());
+        }
+
+        self.validate_matches(&args)?;
+
+        Ok(())
+    }
+
+    // Extracts an argument's value or returns a specific error if the argument is missing.
+    pub fn get_value(&self, arg_name: &'static str) -> Result<String> {
+        self.args
+            .get(arg_name)
+            .ok_or_else(|| Error::UnexpectedArgument(arg_name.to_string()))?
+            .arg_value()
+            .ok_or_else(|| Error::MissingValue(arg_name.to_string()))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn build_parser(arg_parser: ArgParser) -> ArgParser {
```

Comment:
Why receive the parser as an argument here? It seems that this is always called with `ArgParser::default()`.